drivers/counter/counter_nrfx_rtc.c

```diff
@@ -68,226 +74,428 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
-			return -EBUSY;
-		}
+	u32_t max_abs_val;
+	bool absolute = flags & COUNTER_ALARM_CFG_ABSOLUTE;
+	bool trigger_late = flags & COUNTER_ALARM_CFG_EXPIRE_WHEN_LATE;
+
+	__ASSERT(nrf_rtc_int_is_enabled(rtc, int_mask) == 0,
+			"Expected that CC interrupt is disabled.");
+
+	evt = RTC_CHANNEL_EVENT_ADDR(chan);
+	top =  get_dev_data(dev)->top;
+	max_abs_val = top - get_dev_data(dev)->guard_period;
+	now = nrf_rtc_counter_get(rtc);
+
+	/* First take care of a risk of an event coming from CC being set to
+	 * next tick. Reconfigure CC to future (now tick is the furtherest
+	 * future). If CC was set to next tick we need to wait for up to 15us
+	 * (half of 32k tick) and clean potential event. After that time there
+	 * is no risk of unwanted event.
+	 */
+	prev_val = nrf_rtc_cc_get(rtc, chan);
+	nrf_rtc_event_clear(rtc, evt);
+	nrf_rtc_cc_set(rtc, chan, now);
+	nrf_rtc_event_enable(rtc, int_mask);
+
+	if (ticks_sub(dev, prev_val, now, top) == 1) {
+		NRFX_DELAY_US(15);
+		nrf_rtc_event_clear(rtc, evt);
 	}
 
-	nrfx_rtc_cc_disable(rtc, TOP_CH);
 
-	dev_data->top_cb = cfg->callback;
-	dev_data->top_user_data = cfg->user_data;
-	dev_data->top = cfg->ticks;
-	nrfx_rtc_cc_set(rtc, TOP_CH, cfg->ticks, false);
+	/* Special handling required if value is next tick as that will not
+	 * generate event. Set COMPARE to <now> + 2. If meanwhile RTC counter
+	 * got incremented it means it's too late and alarm should expire.
+	 */
+	now = nrf_rtc_counter_get(rtc);
 
-	if (!(cfg->flags & COUNTER_TOP_CFG_DONT_RESET)) {
-		nrfx_rtc_counter_clear(rtc);
-	} else if (counter_nrfx_read(dev) >= cfg->ticks) {
-		err = -ETIME;
-		if (cfg->flags & COUNTER_TOP_CFG_RESET_WHEN_LATE) {
-			nrfx_rtc_counter_clear(rtc);
-		}
+	if (absolute) {
+		val = skip_zero_on_custom_top(val, top);
+	} else {
+		/* Half of top value is taken as an arbitrary value. It is
+		 * assumed that CC might be set too late only if requested
+		 * relative timeout is shorter than half of top value.
+		 */
+		trigger_late = val < (top / 2);
+		/* limit max to detect short relative being set too late. */
+		max_abs_val = trigger_late ? top / 2 : top;
+		val = ticks_add(dev, now, val, top);
 	}
 
-	if (cfg->callback || sw_wrap_required(dev)) {
-		nrfx_rtc_int_enable(rtc, COUNTER_TOP_INT_MASK);
+	diff = ticks_sub(dev, val, now, top);
+	if (diff == 1) {
+		val = ticks_add(dev, val, 1, top);
+		nrf_rtc_cc_set(rtc, chan, val);
+		get_dev_data(dev)->cc_adjusted |= BIT(chan);
+		if (nrf_rtc_counter_get(rtc) != now) {
+			set_cc_int_pending(dev, chan);
+		} else {
+			nrf_rtc_int_enable(rtc, int_mask);
+		}
+	} else {
+		nrf_rtc_cc_set(rtc, chan, val);
+		now = nrf_rtc_counter_get(rtc);
+		diff = ticks_sub(dev, val - 1, now, top);
+		if (diff > max_abs_val) {
+			if (absolute) {
+				err = -ETIME;
+			}
+
+			if (trigger_late) {
+				set_cc_int_pending(dev, chan);
+			} else {
+				get_nrfx_config(dev)->ch_data[chan].callback =
+									NULL;
+			}
+		} else if (diff == 0) {
```

Comment:
This case was already handled in `if (diff == 1)` block [above](https://github.com/zephyrproject-rtos/zephyr/pull/15510/files#diff-cdca186ce3c3dea4b9e41fdc13ff7f2eR243), wasn't it?

Reply:
it's a new diff calculated after setting CC (now is read after setting CC). It maybe that procedure was interrupted and what was 5 ticks in the future, now is 1 tick in the future. We don't know when interruption occurred. It may occurred before setting CC and it was set for COUNTER+1 and CC event will not occur.

Reply:
Right, I missed that. This implementation is not really reviewer-friendly. :wink:
A word of comment on this re-reading of `now`, explaining why it is done, would help a bit probably.

Reply:
added comment, I've also extracted `CC == COUNTER+1` case handling to a function and added description of that function.